### PR TITLE
[VL] Enrich Spark task OOM error message phase 2: include Velox's plan node allocations to OOM message

### PR DESCRIPTION
**Pull request by zhztheplayer**
https://github.com/oap-project/gluten/pull/2866 This is the previous PR (phase 1), which only collects memory usage stats from the root nodes.

This PR collects stats from the children as well.

New OOM message:
```
23/08/28 11:25:51 ERROR ManagedReservationListener: Error reserving memory from target
io.glutenproject.memory.memtarget.ThrowOnOomMemoryTarget$OutOfMemoryException: Not enough spark off-heap execution memory. Acquired: 16777216, granted: 0. Try tweaking config option spark.memory.offHeap.size to get larger space to run this application. 
Current config settings: 
	spark.gluten.memory.offHeap.size.in.bytes=10.0 MiB
	spark.gluten.memory.task.offHeap.size.in.bytes=5.0 MiB
Memory consumer stats:
	GlutenMemoryConsumer/WholeStageIterator@b87fe9c: Current used bytes:    8.0 MiB, peak bytes:    8.0 MiB
	+- result_iterator:                              Current used bytes:    2.0 MiB, peak bytes:    2.0 MiB
	|  \- task.Gluten stage-8 task-8:                Current used bytes:    2.0 MiB, peak bytes:    2.0 MiB
	|     +- node.1:                                 Current used bytes: 1024.0 KiB, peak bytes: 1024.0 KiB
	|     |  \- op.1.0.0.FilterProject:              Current used bytes:   53.0 KiB, peak bytes:   53.0 KiB
	|     \- node.0:                                 Current used bytes: 1024.0 KiB, peak bytes: 1024.0 KiB
	|        +- op.0.0.0.TableScan:                  Current used bytes:   97.4 KiB, peak bytes:   97.8 KiB
	|        \- op.0.0.0.TableScan.test-hive:        Current used bytes:      0.0 B, peak bytes:      0.0 B
	\- WholeStageIterator_default_leaf:              Current used bytes:      0.0 B, peak bytes:      0.0 B
	OverAcquire.DummyTarget@183e9be3:                Current used bytes:      0.0 B, peak bytes:      0.0 B
	GlutenMemoryConsumer/ShuffleWriter@654839d6:     Current used bytes:      0.0 B, peak bytes:      0.0 B
	\- ShuffleWriter_default_leaf:                   Current used bytes:      0.0 B, peak bytes:      0.0 B
```

Also added a new protobuf folder `gluten-core/src/main/resources/io/glutenproject/proto` for swapping structured objects between JNI. In future we can create more proto files for other objects (e.g. config, metrics) to simplify JNI codes.

### CODE REVIEW

**Comment by zhouyuan** [comment]
copy @jackylee-ch

**Comment by zhztheplayer** [comment]
> so this patch can collect the operator level memory stats, is there any difference with this new design?

https://github.com/oap-project/gluten/pull/2866 This is the previous PR (phase 1), which only collects stats from the root nodes.

This PR collects the children as well.

File: gluten-core/src/main/resources/io/glutenproject/proto/memory.proto [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+package gluten;
+
+option java_package = "io.glutenproject.proto";
+option java_multiple_files = true;
+
+message MemoryUsageStats {
+  int64 peak = 1;
+  int64 current = 2;
+  map<string, MemoryUsageStats> children = 3;
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply gluten-core/src/main/resources/io/glutenproject/proto/memory.proto:12]: added

**Comment by liuneng1994** [comment]
Run Gluten Clickhouse CI

**Comment by zuochunwei** [comment]
good job

**Comment by zhztheplayer** [comment]
CI error is not related, probably an environment issue?

```
2023-08-29T06:13:04.9861854Z java.lang.IllegalStateException: Cannot find any build directories.
2023-08-29T06:13:04.9864119Z 	at org.apache.spark.launcher.CommandBuilderUtils.checkState(CommandBuilderUtils.java:228)
2023-08-29T06:13:04.9866886Z 	at org.apache.spark.launcher.AbstractCommandBuilder.getScalaVersion(AbstractCommandBuilder.java:240)
2023-08-29T06:13:04.9869322Z 	at org.apache.spark.launcher.AbstractCommandBuilder.buildClassPath(AbstractCommandBuilder.java:194)
2023-08-29T06:13:04.9871326Z 	at org.apache.spark.launcher.AbstractCommandBuilder.buildJavaCommand(AbstractCommandBuilder.java:117)
2023-08-29T06:13:04.9873234Z 	at org.apache.spark.launcher.WorkerCommandBuilder.buildCommand(WorkerCommandBuilder.scala:39)
2023-08-29T06:13:04.9875029Z 	at org.apache.spark.launcher.WorkerCommandBuilder.buildCommand(WorkerCommandBuilder.scala:45)
2023-08-29T06:13:04.9876678Z 	at org.apache.spark.deploy.worker.CommandUtils$.buildCommandSeq(CommandUtils.scala:63)
2023-08-29T06:13:04.9878205Z 	at org.apache.spark.deploy.worker.CommandUtils$.buildProcessBuilder(CommandUtils.scala:51)
2023-08-29T06:13:04.9880021Z 	at org.apache.spark.deploy.worker.ExecutorRunner.org$apache$spark$deploy$worker$ExecutorRunner$$fetchAndRunExecutor(ExecutorRunner.scala:158)
2023-08-29T06:13:04.9881828Z 	at org.apache.spark.deploy.worker.ExecutorRunner$$anon$1.run(ExecutorRunner.scala:79)
2023-08-29T06:13:04.9883272Z 14:13:04.984 ERROR org.apache.spark.deploy.worker.ExecutorRunner: Error running executor
```

@zhouyuan @xieqi

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by zhztheplayer** [comment]
Thanks for reviewing!

File: cpp/velox/memory/VeloxMemoryManager.cc [hunk 2]
```diff
@@ -199,4 +199,26 @@ VeloxMemoryManager::VeloxMemoryManager(
...
+  pool->visitChildren([&](velox::memory::MemoryPool* pool) -> bool {
+    stats.mutable_children()->emplace(pool->name(), collectMemoryUsageStatsInternal(pool));
+    return true;
+  });
+  return stats;
+}
+} // namespace
+
+const MemoryUsageStats VeloxMemoryManager::collectMemoryUsageStats() const {
+  return collectMemoryUsageStatsInternal(veloxPool_.get());
```
Comment (Yohahaha) [inline cpp/velox/memory/VeloxMemoryManager.cc:217]: veloxPool_ has been renamed.

Comment (ulysses-you) [inline cpp/velox/memory/VeloxMemoryManager.cc:217]: there is no `veloxPool_`, do you mean `veloxLeafPool_` ?

**Comment by zhztheplayer** [comment]
@Yohahaha @ulysses-you Working on a hotfix. Thanks for reporting.

**Comment by zhouyuan** [comment]
merge conflict w/ 
https://github.com/oap-project/gluten/pull/2939

File: cpp/core/CMakeLists.txt [hunk 3]
```diff
@@ -250,22 +262,34 @@ else()
...
+        --proto_path
+        ${SUBSTRAIT_PROTO_SRC_DIR}/
+        --cpp_out
+        ${PROTO_OUTPUT_DIR}
+        ${SUBSTRAIT_PROTO_FILES}
+        DEPENDS  ${SUBSTRAIT_PROTO_DIR}
+        COMMENT "Running Substrait PROTO compiler"
+        VERBATIM)
+
+add_custom_command(OUTPUT ${GLUTEN_PROTO_OUTPUT_FILES}
```
Comment (Yohahaha) [inline cpp/core/CMakeLists.txt:276]:
```
CMake Error at core/CMakeLists.txt:276 (add_custom_command):
  add_custom_command Wrong syntax.  A TARGET or OUTPUT must be specified.
```

CMake error in my build machine.

**Pull request merged**